Add ComplexityConfiguration functionality and ignore the introspectio…

bourgoisa opened the pull request:
We would like to have the complexityConfiguration from graphql.net in the conventions library. Which we have worked out in this pull request. Can you please check this and inform us of your findings

Commit:
Add ComplexityConfiguration functionality and ignore the introspectionquery

tlil reviewed (commented):
Thanks, @bourgoisa – I had this on my todo-list.

This LGTM - only thing, you might want to add it to the `RequestHandlerBuilder` as well. ([see here](https://github.com/graphql-dotnet/conventions/blob/master/src/GraphQL.Conventions/Web/RequestHandler.cs#L18))

huysentruitw commented on src/GraphQL.Conventions/Adapters/Engine/GraphQLEngine.cs:
```diff
@@ -213,15 +214,23 @@ internal IDependencyInjector DependencyInjector
                 rules = new[] { new NoopValidationRule() };
             }
+
+            complexityConfiguration = (query.ToLower().Contains("introspectionquery") &&
```
Complexity validation can now simply be bypassed by a query like this:

```
query introspectionquery__schema {
	someQuery {
    someField
  }
}
```

Since the complexity configuration needs to be passed for each `Execute` (each query), I would leave that decision/detection up to the external calling code and remove this condition here.

In short, if the calling code wants to bypass the introspection query, then they are responsible for detecting introspection and not call `WithComplexityConfiguration` in that particular case.

tlil replied:
👍  Agreed

tlil replied:
@bourgoisa thanks for the update - could you please make the changes that @huysentruitw proposed as well, and I'll get this merged?

Commit:
CSFOUR-4283: added complexityConfiguration to requesthandler

huysentruitw commented on test/Tests/Web/RequestHandlerTests.cs:
```diff
@@ -40,6 +41,23 @@ public class RequestHandlerTests : TestBase
+            response.Warnings.Count.ShouldEqual(0);
+        }
+        
         class TestQuery
```
Maybe also add an additional test where the query is too complex according to given `MaxDepth`. Only then you're sure that the `ComplexityConfiguration` is actually being used.

bourgoisa commented:
Now I removed the check for the introspectionquery as discussed above.

Also now I have added  two unit tests to test the complexityconfiguration added on the Executor.
If any remarks please let me know.

Commit:
Removed the check for introspectionquery concerning the complexityConfiguration

Commit:
Added unit tests for ComplexityConfiguration and add run stylingprocessor on unit tests

huysentruitw committed:
Review

huysentruitw committed:
Undo unrelated line-addition

huysentruitw commented:
Also added one additional unit test and did some small modifications to the PR. I think we're at it now :)

tlil commented:
Thanks guys. I'm pushing out v1.2.8 here shortly.
